index.d.ts

```diff
@@ -15,55 +13,53 @@ export type LocaleData = {
   }
 };
 
+type Normalize<Input extends string | number> = Input extends string ? Lowercase<Input> : Input;
 
 
 export function registerLocale(localeData: LocaleData): void;
-export function alpha2ToAlpha3(alpha2: string | Alpha2Code): string;
-export function alpha2ToNumeric(alpha2: string | Alpha2Code): string;
-export function alpha3ToAlpha2(alpha3: string | Alpha3Code): string;
-export function alpha3ToNumeric(alpha3: string | Alpha3Code): string;
-export function numericToAlpha2(numeric: number | string): string;
-export function numericToAlpha3(numeric: number | string): string;
+export function alpha2ToAlpha3<Input extends string>(alpha2: Input): Normalize<Input> extends Normalize<Alpha2Code> ? Alpha3Code : Alpha3Code | undefined;
+export function alpha2ToNumeric<Input extends string>(alpha2: Input): Normalize<Input> extends Normalize<Alpha2Code> ? StringNumericCode : StringNumericCode | undefined;
+export function alpha3ToAlpha2<Input extends string>(alpha3: Input): Normalize<Input> extends Normalize<Alpha3Code> ? Alpha2Code : Alpha2Code | undefined;
+export function alpha3ToNumeric<Input extends string>(alpha3: Input): Normalize<Input> extends Normalize<Alpha3Code> ? StringNumericCode : StringNumericCode | undefined;
+export function numericToAlpha2<Input extends string | number>(numeric: Input): Input extends NumericCode ? Alpha2Code : Alpha2Code | undefined;
+export function numericToAlpha3<Input extends string | number>(numeric: Input): Input extends NumericCode ? Alpha3Code : Alpha3Code | undefined;
 
 /**
  * Returns object map where key is alpha 2 code and value is alpha 3 code
  */
-export function getAlpha2Codes(): { [alpha2Key: string]: string };
+export function getAlpha2Codes(): Record<Alpha2Code, Alpha3Code>;
 
 /**
  * Returns object map where key is alpha 3 code and value is alpha 2 code
  */
-export function getAlpha3Codes(): { [alpha3Key: string]: string };
+export function getAlpha3Codes(): Record<Alpha3Code, Alpha2Code>;
 
 /**
  * Returns object map where key is numeric code and value is alpha 2 code
  */
-export function getNumericCodes(): { [numericKey: string]: string };
+export function getNumericCodes(): Record<StringNumericCode, Alpha2Code>;
 
 /**
  * @param countryCode  Alpha2 or Alpha3 or Numeric
  * @param lang        ISO 639-1 format string
  */
-export function getName(
-  countryCode: string | number | Alpha2Code | Alpha3Code,
-  lang: string
-): string;
+export function getName(countryCode: string | number, lang: string): string | undefined;
```

Comment:
We can't have a more specific type here since A) the name of the country varies arbitrarily by language; and B) if the language is bad, then it'll return undefined; I think it's a bad idea to try to add all the possible language codes, it's overkill.